pkg/cache/internal/factory.go

```diff
@@ -0,0 +1,212 @@
+	resync time.Duration) *InformersMap {
+	ip := &InformersMap{
+		config:         config,
+		Scheme:         scheme,
+		mapper:         mapper,
+		informersByGVK: make(map[schema.GroupVersionKind]*MapEntry),
+		codecs:         serializer.NewCodecFactory(scheme),
+		paramCodec:     runtime.NewParameterCodec(scheme),
+		resync:         resync,
+	}
+	return ip
+}
+
+// MapEntry contains the cached data for an Informer
+type MapEntry struct {
+	// Informer is the cached informer
+	Informer cache.SharedIndexInformer
+
+	// Reader wraps Informer and implements the Reader interface for a single type
+	Reader Reader
+}
+
+// InformersMap create and caches Informers for (runtime.Object, schema.GroupVersionKind) pairs.
+//It uses a standard parameter codec constructed based on the given generated Scheme.
+type InformersMap struct {
+	// Scheme maps runtime.Objects to GroupVersionKinds
+	Scheme *runtime.Scheme
+
+	// config is used to talk to the apiserver
+	config *rest.Config
+
+	// mapper maps GroupVersionKinds to Resources
+	mapper meta.RESTMapper
+
+	// informersByGVK is the cache of informers keyed by groupVersionKind
+	informersByGVK map[schema.GroupVersionKind]*MapEntry
+
+	// codecs is used to create a new REST client
+	codecs serializer.CodecFactory
+
+	// paramCodec is used by list and watch
+	paramCodec runtime.ParameterCodec
+
+	// stop is the stop channel to stop informers
+	stop <-chan struct{}
+
+	// resync is the frequency the informers are resynced
+	resync time.Duration
+
+	// mu guards access to the map
+	mu sync.Mutex
+
+	// start is true if the informers have been started
+	started bool
+}
+
+// Start calls Run on each of the informers and sets started to true.  Blocks on the stop channel.
+func (ip *InformersMap) Start(stop <-chan struct{}) error {
+	func() {
+		ip.mu.Lock()
+		defer ip.mu.Unlock()
+
+		// Set the stop channel so it can be passed to informers that are added later
+		ip.stop = stop
+
+		// Start each informer
+		for _, informer := range ip.informersByGVK {
+			go informer.Informer.Run(stop)
+		}
+
+		// Set started to true so we immediately start any informers added later.
+		ip.started = true
+	}()
+	<-stop
+	return nil
+}
+
+// WaitForCacheSync waits until all the caches have been synced
+func (ip *InformersMap) WaitForCacheSync(stop <-chan struct{}) bool {
+	syncedFuncs := make([]cache.InformerSynced, 0, len(ip.informersByGVK))
+	for _, informer := range ip.informersByGVK {
+		syncedFuncs = append(syncedFuncs, informer.Informer.HasSynced)
+	}
+	return cache.WaitForCacheSync(stop, syncedFuncs...)
+}
+
+// Get will create a new Informer and added it to the map of InformersMap if none exists.  Returns
+// the Informer from the map.
+func (ip *InformersMap) Get(gvk schema.GroupVersionKind, obj runtime.Object) (*MapEntry, error) {
+	// Do the mutex part in its own function so we can use defer without blocking pieces that don't
+	// to be locked
+	var sync bool
+	i, err := func() (*MapEntry, error) {
+		ip.mu.Lock()
```

Comment:
This is going to cause issues, because it means that only one read operation can happen across any index at once, which is suboptimal. That's one of the reasons why the separation between creation and client lookup was useful.  I think what we might want to do is

```
// assume the happy path first
with read-only-lock {
  if i, ok := informers[gvk]; ok {
    return i
  }
}

// sad path, first time only
with write-lock {
    // check again in case we raced
    if i, ok := informers[gvk]; ok {
    return i
  }

  // only construct if we actually need to do so
  i := construct informer
  informers[gvk] = i
  return i
}
```

Reply:
Done